crates/refunder/src/refund_service.rs

```diff
@@ -1,47 +1,138 @@
+
 impl RefundService {
-    pub fn new(db: PgPool, min_validity_duration: i64, min_slippage_in_bps: u64) -> Self {
+    pub fn new(
+        db: PgPool,
+        web3: Web3,
+        ethflow_contract: CoWSwapEthFlow,
+        min_validity_duration: i64,
+        min_slippage_bps: u64,
+    ) -> Self {
         RefundService {
             db,
+            web3,
+            ethflow_contract,
             min_validity_duration,
-            min_slippage: min_slippage_in_bps as f64 / 10000f64,
+            min_slippage: min_slippage_bps as f64 / 10000f64,
         }
     }
-
     pub async fn try_to_refund_all_eligble_orders(&self) -> Result<()> {
-        // Step 1:
-        // Look for all refundable orders by calling the database
+        let refundable_order_uids = self.get_refundable_ethflow_orders_from_db().await?;
+
+        let (refunded, _to_be_refunded) = self
+            .identify_already_refunded_order_uids_via_web3_calls(refundable_order_uids)
+            .await?;
+
+        self.update_already_refunded_orders_in_db(refunded).await?;
+
+        // Send out tx to deleteOrders, and wait for 5 block for the tx
+        // to be mined
+        // todo()
+
+        Ok(())
+    }
+
+    async fn update_already_refunded_orders_in_db(
+        &self,
+        refunded: Vec<(OrderUid, bool)>,
+    ) -> Result<()> {
+        let refunded_uids: Vec<OrderUid> = refunded.into_iter().map(|(uid, _)| uid).collect();
+
+        let mut transaction = self.db.begin().await?;
+        mark_eth_orders_as_refunded(&mut transaction, refunded_uids.as_slice())
+            .await
+            .map_err(|err| {
+                anyhow!(
+                    "Error while retrieving updating the already refunded orders:{:?}",
+                    err
+                )
+            })?;
+
+        Ok(transaction.commit().await?)
+    }
+
+    async fn get_refundable_ethflow_orders_from_db(&self) -> Result<Vec<EthOrderPlacement>> {
         let unix_timestamp = SystemTime::now()
             .duration_since(SystemTime::UNIX_EPOCH)
             .unwrap()
             .as_secs() as i64;
         let mut ex = self.db.acquire().await?;
-        let _refundable_order_uids = refundable_orders(
+        refundable_orders(
             &mut ex,
             unix_timestamp,
             self.min_validity_duration,
             self.min_slippage,
         )
-        .await?;
+        .await
+        .map_err(|err| {
+            anyhow!(
+                "Error while retrieving the refundable ethflow orders from db: {:?}",
+                err
+            )
+        })
+    }
 
-        // Step 2:
-        // take out orderUids that are already refunded
-        // by making a batched web3 call. Update refunded ones.
+    async fn identify_already_refunded_order_uids_via_web3_calls(
+        &self,
+        refundable_order_uids: Vec<EthOrderPlacement>,
+    ) -> Result<SplittedOrderUids> {
+        let mut batch = Web3CallBatch::new(self.web3.transport().clone());
```

Comment:
Depending on which PR merges first you might not have to use the `Web3CallBatch` thing since Nick [refactors](https://github.com/cowprotocol/services/pull/696) the code base to use a batching mechanism by default.